File: python/src/agent_framework/core/context_engine.py
from __future__ import annotations

from pathlib import Path

from agent_framework.core.contracts import ContextEngine
from agent_framework.core.domain import AgentMode, Message


DEFAULT_CONTEXT_FILES = ["AGENTS.md", "SOUL.md", "TOOLS.md"]


class WorkspaceContextEngine(ContextEngine):
    def __init__(self, workspace_root: str, context_files: list[str] | None = None) -> None:
        self._workspace_root = Path(workspace_root)
        self._context_files = context_files or DEFAULT_CONTEXT_FILES

    async def build(self, session_id: str, mode: AgentMode, history: list[Message]) -> tuple[str, list[Message]]:
        del session_id
        sections = [_mode_prompt(mode)]
        for file_name in self._context_files:
            content = _read_optional(self._workspace_root / file_name)
            if content:
                sections.append(f"## {file_name}\n{content}")

        skills_dir = self._workspace_root / "skills"
        if skills_dir.exists():
            skill_sections = []
            for skill_dir in sorted(skills_dir.iterdir()):
                skill_file = skill_dir / "SKILL.md"
                if skill_file.exists():
                    skill_sections.append(f"### Skill: {skill_dir.name}\n{skill_file.read_text(encoding='utf-8')}")
            if skill_sections:
                sections.append("## Skills\n" + "\n\n".join(skill_sections))

        system_prompt = "\n\n".join(sections)
        messages = [Message(role="system", content=system_prompt), *[
            message for message in history if message.role != "system"
        ]]
        return system_prompt, messages


def _mode_prompt(mode: AgentMode) -> str:
    if mode == "plan":
        return "You are in plan mode. Produce detailed implementation plans before making changes."
    if mode == "ask":
        return "You are in ask mode. Answer questions without modifying files or running destructive commands."
    return "You are an autonomous coding agent. Use tools to inspect and modify the workspace safely."


def _read_optional(path: Path) -> str | None:
    try:
        return path.read_text(encoding="utf-8")
    except OSError:
        return None

File: typescript/packages/core/src/context-engine.ts
import { readFile, readdir } from 'node:fs/promises';
import { join } from 'node:path';
import type { ContextBuildRequest, ContextBuildResult, ContextEngine } from './types/contracts.js';
import type { AgentMode, Message } from './types/domain.js';

export interface WorkspaceContextOptions {
  workspaceRoot: string;
  contextFiles?: string[];
  skillsDir?: string;
}

const DEFAULT_CONTEXT_FILES = ['AGENTS.md', 'SOUL.md', 'TOOLS.md'];

export class WorkspaceContextEngine implements ContextEngine {
  constructor(private readonly options: WorkspaceContextOptions) {}

  async build(request: ContextBuildRequest): Promise<ContextBuildResult> {
    const sections: string[] = [];
    const modeInstruction = modePrompt(request.mode ?? 'agent');
    sections.push(modeInstruction);

    for (const fileName of this.options.contextFiles ?? DEFAULT_CONTEXT_FILES) {
      const content = await readOptional(join(this.options.workspaceRoot, fileName));
      if (content) {
        sections.push(`## ${fileName}\n${content}`);
      }
    }

    const skills = await this.loadSkills();
    if (skills.length > 0) {
      sections.push(`## Skills\n${skills.join('\n\n')}`);
    }

  const systemPrompt = sections.join('\n\n');
    return {
      systemPrompt,
      messages: prependSystem(request.history, systemPrompt),
    };
  }

  private async loadSkills(): Promise<string[]> {
    const skillsDir = this.options.skillsDir ?? join(this.options.workspaceRoot, 'skills');
    try {
      const entries = await readdir(skillsDir, { withFileTypes: true });
      const skills: string[] = [];
      for (const entry of entries.sort((a, b) => a.name.localeCompare(b.name))) {
        if (entry.isDirectory()) {
          const skillMd = await readOptional(join(skillsDir, entry.name, 'SKILL.md'));
          if (skillMd) {
            skills.push(`### Skill: ${entry.name}\n${skillMd}`);
          }
        }
      }
      return skills;
    } catch {
      return [];
    }
  }
}

function modePrompt(mode: AgentMode): string {
  switch (mode) {
    case 'plan':
      return 'You are in plan mode. Produce detailed implementation plans before making changes.';
    case 'ask':
      return 'You are in ask mode. Answer questions without modifying files or running destructive commands.';
    default:
      return 'You are an autonomous coding agent. Use tools to inspect and modify the workspace safely.';
  }
}

function prependSystem(history: Message[], systemPrompt: string): Message[] {
  const withoutSystem = history.filter((message) => message.role !== 'system');
  return [{ role: 'system', content: systemPrompt }, ...withoutSystem];
}

async function readOptional(path: string): Promise<string | null> {
  try {
    return await readFile(path, 'utf8');
  } catch {
    return null;
  }
}
